sphinx/pycode/function.py

```diff
@@ -0,0 +1,124 @@
+"""
+    sphinx.pycode.function
+    ~~~~~~~~~~~~~~~~~~~~~~
+
+    Utilities parsing function definitions
+
+    :copyright: Copyright 2007-2020 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+import ast
+import sys
+from collections import OrderedDict
+from inspect import Parameter
+from typing import cast
+from typing import Mapping
+
+if sys.version_info > (3, 8):
+    from functools import cached_property
+else:
+    from cached_property import cached_property
+
+
+def stringify(node: ast.AST) -> str:
+    """Stringify an AST node."""
+    if node is None:
+        return None
+    elif isinstance(node, ast.Attribute):
+        return '%s.%s' % (stringify(node.value), node.attr)
+    elif isinstance(node, ast.Bytes):
+        return repr(node.s)
+    elif isinstance(node, ast.Call):
+        args = ([stringify(e) for e in node.args] +
+                ['%s=%s' % (k.arg, stringify(k.value)) for k in node.keywords])
+        return '%s(%s)' % (stringify(node.func), ', '.join(args))
+    elif isinstance(node, ast.Dict):
+        keys = (stringify(k) for k in node.keys)
+        values = (stringify(v) for v in node.values)
+        items = (k + ': ' + v for k, v in zip(keys, values))
+        return '{' + ', '.join(items) + '}'
+    elif isinstance(node, ast.Ellipsis):
+        return '...'
+    elif sys.version_info > (3, 6) and isinstance(node, ast.Constant):
+        return repr(node.value)
+    elif isinstance(node, ast.Index):
+        return stringify(node.value)
+    elif isinstance(node, ast.Lambda):
+        return '<function <lambda>>'  # TODO
+    elif isinstance(node, ast.List):
+        return '[' + ', '.join(stringify(e) for e in node.elts) + ']'
+    elif isinstance(node, ast.Name):
+        return node.id
+    elif isinstance(node, ast.NameConstant):
+        return repr(node.value)
+    elif isinstance(node, ast.Num):
+        return repr(node.n)
+    elif isinstance(node, ast.Set):
+        return "{" + ', '.join(stringify(e) for e in node.elts) + "}"
+    elif isinstance(node, ast.Str):
+        return repr(node.s)
+    elif isinstance(node, ast.Subscript):
+        return "%s[%s]" % (stringify(node.value), stringify(node.slice))
+    elif isinstance(node, ast.Tuple):
+        return ', '.join(stringify(e) for e in node.elts)
+    else:
+        # TODO: f-string
+        raise NotImplementedError('Unable to render %s instance' % type(node).__name__)
+
+
+class Signature:
```

Comment:
I don't think you need a new class here. Why not
```
def signature_from_str(s):
    ...
    return inspect.Signature(parameters=..., return_annotation=...)
```

Reply:
OMG. I mistook `inspect.signature` (function) and inspect.Signature` (class). As you said, we can use `inspect.Signature` class directly.

Reply:
done.